File: first_time/greedy/860.柠檬水找零.py
#
# @lc app=leetcode.cn id=860 lang=python3
#
# [860] 柠檬水找零
#

# @lc code=start
class Solution:
    def lemonadeChange(self, bills):
        dollars = {5:0,10:0,20:0}
        for bill in bills:
            if bill == 5:
                dollars[5] += 1
            elif bill == 10:
                if dollars[5] == 0:
                    return False
                else:
                    dollars[5] -= 1
                    dollars[10] += 1
            elif bill ==20:
                if dollars[10] > 0 and dollars[5] > 0:
                    dollars[10] -= 1
                    dollars[5] -= 1
                    dollars[20] += 1
                elif dollars[5] >=3:
                    dollars[5] -= 3
                    dollars[20] += 1
                else:
                    return False
            else:
                pass
        return True
                
                
            

        
# @lc code=end

File: second_time/deque/860.柠檬水找零.cpp
/*
 * @lc app=leetcode.cn id=860 lang=cpp
 *
 * [860] 柠檬水找零
 */

// @lc code=start
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        int five = 0, ten = 0, twenty = 0;
        for(int i : bills){
            if (i==5) five ++;
            else if (i==10) five --, ten ++;
            else if (ten > 0) ten --, five--, twenty++;
            else five -= 3, twenty++;
            if (five < 0) return false;
        }
        return true;
    }
};
// @lc code=end

